source-genesys/source_genesys/api.py

```diff
@@ -0,0 +1,185 @@
+import asyncio
+
+from datetime import datetime, timedelta, UTC
+from logging import Logger
+from typing import AsyncGenerator, Any
+from estuary_cdk.capture.common import LogCursor, PageCursor
+from estuary_cdk.http import HTTPSession, RateLimiter
+
+from .models import (
+    User,
+    UserResponse,
+    Conversation,
+    CreateJobResponse,
+    CheckJobStatusResponse,
+    JobResultsResponse,
+)
+
+COMMON_API = "https://api"
+DATETIME_STRING_FORMAT = "%Y-%m-%dT%H:%M:%SZ"
+
+
+def _dt_to_str(dt: datetime) -> str:
+    return dt.strftime(DATETIME_STRING_FORMAT)
+
+def _str_to_dt(date: str) -> datetime:
+    return datetime.fromisoformat(date)
+
+
+async def snapshot_users(
+        http: HTTPSession,
+        domain: str,
+        log: Logger,
+) -> AsyncGenerator[User, None]:
+    """
+    API docs - https://developer.genesys.cloud/useragentman/users/#get-api-v2-users
+    """
+    url = f"{COMMON_API}.{domain}/api/v2/users"
+
+    params = {
+        "pageSize": 500,
+        "pageNumber": 1,
+        "sortOrder": "ASC",
+    }
+
+    lastPageNumber = 1
+
+    while params["pageNumber"] <= lastPageNumber:
+        response = UserResponse.model_validate_json(
+            await http.request(log, url, params=params)
+        )
+
+        users = response.entities
+        for user in users:
+            yield user
+
+        params["pageNumber"] += 1
+        lastPageNumber = response.pageCount
+
+
+async def fetch_conversations(
+        http: HTTPSession,
+        domain: str,
+        log: Logger,
+        log_cursor: LogCursor,
+) -> AsyncGenerator[Conversation | LogCursor, None]:
+    """
+    Submits an asynchronous job for all conversations
+    that have ended between start_date and the present.
+    """
+    assert isinstance(log_cursor, datetime)
+
+    updated_cursor = log_cursor
+
+    async for record in _perform_conversation_job(
+        http, domain, log, log_cursor
+    ): 
+        # Async analytics jobs return any conversations that started or ended within the requested date window.
+        # The stream's cursor should be updated to the record's most recent datetime.
+        most_recent_conversation_dt = record.conversationEnd or record.conversationStart
+        if most_recent_conversation_dt > log_cursor:
+            updated_cursor = max(updated_cursor, most_recent_conversation_dt)
+            yield record
```

Comment:
Good point. It's slightly tricky since the results contain records that have either started _or_ ended within the requested date window. For example, I'm pretty sure this is a valid result set we could receive:
```python
# Job requested results between 2024-11-08 and 2024-11-11.
# Records are abbreviated for clarity.

results = [
	{"id": 0, "start": "2024-11-01T00:00:00Z", "end": "2024-11-11T00:00:00Z"},
	{"id": 1, "start": "2024-11-08T00:00:00Z", "end": "2024-11-11T00:00:00Z"},
	{"id": 2, "start": "2024-11-08T00:00:00Z"},
	{"id": 3, "start": "2024-11-09T00:00:00Z", "end": "2024-11-10T00:00:00Z"},
	{"id": 4, "start": "2024-11-09T01:00:00Z", "end": "2024-11-09T23:00:00Z"}
]
```


I can think of a couple approaches for checkpointing:
1. Date slicing - only query for X days of data in a single job instead of an unbounded amount. Should work, but would slow us down since we'd have to submit multiple different jobs.
2. Checkpoint the previous value of `updated_cursor` after we see that `updated_cursor` moves forward in time, guaranteeing that we've yielded all the records at the previous `updated_cursor` value. We could also only do that checkpoint after a certain # of documents have been yielded. Implementation could be something like:
```python
    most_recent_end = log_cursor
    updated_cursor = log_cursor
    previous_updated_cursor = updated_cursor
    count = 0
    MIN_CHECKPOINT_INTERVAL = 1000

    async for record in _perform_conversation_job(
        http, domain, log, log_cursor
    ): 
        # Async analytics jobs return any conversations that started or ended within the requested date window.
        start = record.conversationStart
        end = record.conversationEnd
        # Yield records that started before/at the log_cursor but have ended at/afterwards.
        if start <= log_cursor and end is not None:
            yield record
            most_recent_end = max(most_recent_end, end)
        # Yield records that started after the previous cursor
        elif start > log_cursor:
            yield record
            count += 1
            updated_cursor = start

            # Only checkpoint if most_recent_start has changed (i.e. we've yielded all records for a
            # specific start datetime) AND we've emitted more records than our MIN_CHECKPOINT_INTERVAL.
            if updated_cursor != previous_updated_cursor and count >= MIN_CHECKPOINT_INTERVAL:
                yield previous_updated_cursor
                count = 0

            previous_updated_cursor = updated_cursor

    # Yield the most recent datetime, either the most recent start or the most recent end from the records that started before log_cursor.
    updated_cursor = max(updated_cursor, most_recent_end)
    if updated_cursor > log_cursor:
        yield updated_cursor
```

I'm thinking option 2 is nicer since it only submits a single job for Genesys to complete, plus we'd likely need some of that cursor management logic even if we did date window slicing instead. 

Does going with option 2 sound good to you?

Reply:
> Good point. It's slightly tricky since the results contain records that have either started _or_ ended within the requested date window. For example, I'm pretty sure this is a valid result set we could receive:
> 
> ```python
> # Job requested results between 2024-11-08 and 2024-11-11.
> # Records are abbreviated for clarity.
> 
> results = [
> 	{"id": 0, "start": "2024-11-01T00:00:00Z", "end": "2024-11-11T00:00:00Z"},
> 	{"id": 1, "start": "2024-11-08T00:00:00Z", "end": "2024-11-11T00:00:00Z"},
> 	{"id": 2, "start": "2024-11-08T00:00:00Z"},
> 	{"id": 3, "start": "2024-11-09T00:00:00Z", "end": "2024-11-10T00:00:00Z"},
> 	{"id": 4, "start": "2024-11-09T01:00:00Z", "end": "2024-11-09T23:00:00Z"}
> ]
> ```
> 
> I can think of a couple approaches for checkpointing:
> 
> 1. Date slicing - only query for X days of data in a single job instead of an unbounded amount. Should work, but would slow us down since we'd have to submit multiple different jobs.
> 2. Checkpoint the previous value of `updated_cursor` after we see that `updated_cursor` moves forward in time, guaranteeing that we've yielded all the records at the previous `updated_cursor` value. We could also only do that checkpoint after a certain # of documents have been yielded. Implementation could be something like:
> 
> ```python
>     most_recent_end = log_cursor
>     updated_cursor = log_cursor
>     previous_updated_cursor = updated_cursor
>     count = 0
>     MIN_CHECKPOINT_INTERVAL = 1000
> 
>     async for record in _perform_conversation_job(
>         http, domain, log, log_cursor
>     ): 
>         # Async analytics jobs return any conversations that started or ended within the requested date window.
>         start = record.conversationStart
>         end = record.conversationEnd
>         # Yield records that started before/at the log_cursor but have ended at/afterwards.
>         if start <= log_cursor and end is not None:
>             yield record
>             most_recent_end = max(most_recent_end, end)
>         # Yield records that started after the previous cursor
>         elif start > log_cursor:
>             yield record
>             count += 1
>             updated_cursor = start
> 
>             # Only checkpoint if most_recent_start has changed (i.e. we've yielded all records for a
>             # specific start datetime) AND we've emitted more records than our MIN_CHECKPOINT_INTERVAL.
>             if updated_cursor != previous_updated_cursor and count >= MIN_CHECKPOINT_INTERVAL:
>                 yield previous_updated_cursor
>                 count = 0
> 
>             previous_updated_cursor = updated_cursor
> 
>     # Yield the most recent datetime, either the most recent start or the most recent end from the records that started before log_cursor.
>     updated_cursor = max(updated_cursor, most_recent_end)
>     if updated_cursor > log_cursor:
>         yield updated_cursor
> ```
> 
> I'm thinking option 2 is nicer since it only submits a single job for Genesys to complete, plus we'd likely need some of that cursor management logic even if we did date window slicing instead.
> 
> Does going with option 2 sound good to you?

Hmm I see the complexity here. I think we should actually just leave the implementation like you have it, being a simple single-shot with no intermediate checkpoint, and revisit once we see some real production use and if this is a problem at all.